diff --git a/modules/contrib/taxonomy_orphanage/taxonomy_orphanage.batch.inc b/modules/contrib/taxonomy_orphanage/taxonomy_orphanage.batch.inc
index ab136b6..d52c314 100644
--- a/modules/contrib/taxonomy_orphanage/taxonomy_orphanage.batch.inc
+++ b/modules/contrib/taxonomy_orphanage/taxonomy_orphanage.batch.inc
@@ -39,10 +39,14 @@
   $res = $query->execute();
   foreach ($res as $target) {
     $context['sandbox']['progress']++;
-    $entity = array_shift(entity_load($target->entity_type, array($target->entity_id)));
+    $loaded_entity = entity_load($target->entity_type, array($target->entity_id));
+    $entity = array_shift($loaded_entity);
     if (!$entity) {
       continue;
     }
+    if (empty($entity->language)) {
+      $entity->language = LANGUAGE_NONE;
+    }
     foreach ($entity->{$field['field_name']}[$entity->language] as $k => $term) {
       if ($term['tid'] == $target->{$field['field_name'] . '_tid'}) {
         unset($entity->{$field['field_name']}[$entity->language][$k]);