diff --git a/views/draggableviews_join_handler.inc b/views/draggableviews_join_handler.inc
index dd7c0c46c..aa4d66ee9 100644
--- a/views/draggableviews_join_handler.inc
+++ b/views/draggableviews_join_handler.inc
@@ -54,7 +54,7 @@

       // Arguments passed to view (including exposed filters).
       $view_arguments = isset($view->args) ? $view->args : array();
-      if (isset($view->exposed_raw_input)) {
+      if (!empty($view->exposed_raw_input)) {
         $view_arguments += $view->exposed_raw_input;
         // Remove 'items_per_page' argument if added by exposed pager filter.
         unset($view_arguments['items_per_page']);
