diff --git a/geoip.install b/geoip.install
index 43a3074..05c6b2e 100644
--- a/geoip.install
+++ b/geoip.install
@@ -50,7 +50,7 @@ function geoip_requirements($phase) {
   if (!($file = variable_get('geoip_data_file')) || !file_exists($file)) {
     $requirements['geoip_database']['value'] = l($t('Missing'), 'admin/settings/geoip');
     $requirements['geoip_database']['description'] = $t('The GeoIP database file is missing or not configured. Download the latest file at <a href="@url">Maxmind.com</a>.', array('@url' => $database_link));
-    $requirements['geoip_database']['severity'] = ($phase == 'runtime') ? REQUIREMENT_ERROR : REQUIREMENT_ERROR;
+    $requirements['geoip_database']['severity'] = ($phase == 'runtime') ? REQUIREMENT_ERROR : REQUIREMENT_INFO;
   }
   else {
     $mtime = filemtime($file);
